File: Solutions/77-combinations/combinations.cs
public class Solution {
    public IList<IList<int>> Combine(int n, int k) {
        // This will store all valid combinations
        var result = new List<IList<int>>();
        // Temporary list to build each combination
        var combination = new List<int>();

        // Start the backtracking process from number 1
        Backtrack(1, n, k, combination, result);

        return result;
    }

    private void Backtrack(int start, int n, int k, List<int> combination, List<IList<int>> result) {
        // Base case: if the current combination has k numbers, it's complete
        if (combination.Count == k) {
            // Add a copy of the current combination to the result
            result.Add(new List<int>(combination));
            return;
        }

        // Loop through numbers from 'start' to 'n'
        // 'start' ensures we only move forward to avoid duplicates
        for (int i = start; i <= n; i++) {
            // Choose the current number by adding it to the combination
            combination.Add(i);
            
            // Explore further by recursively adding the next numbers
            // i + 1 ensures we don't reuse the same number
            Backtrack(i + 1, n, k, combination, result);

            // Un-choose: remove the last number to try the next one
            combination.RemoveAt(combination.Count - 1);
        }
    }
}

File: Solutions/77-combinations/combinations.dart
class Solution {
  List<List<int>> combine(int n, int k) {
    // This will store all the combinations
    List<List<int>> result = [];

    // Helper function for backtracking
    void backtrack(int start, List<int> current) {
      // Base case: if the current combination has k numbers, add it to result
      if (current.length == k) {
        result.add(List.from(current)); // Make a copy of current
        return;
      }

      // Try all possible numbers starting from 'start' to 'n'
      for (int i = start; i <= n; i++) {
        current.add(i); // Choose number i
        backtrack(i + 1, current); // Recurse with next numbers
        current.removeLast(); // Backtrack: remove i and try next
      }
    }

    // Start backtracking from 1 with an empty combination
    backtrack(1, []);
    
    return result;
  }
}
